### PR TITLE
Update gitlab API to version v4 #2068

### PR DESCRIPTION
**Pull request by vanadium23**
Hello, fellows.

I've moved drone gitlab client from api/v3 to api/v4, also slightly change unit test so they've been passing, but imo we need to refactor them in the way other clients do.
I've tested it against local gitlab and gitlab.com and both works, except commit status is not changing, but it's broken from gitlab API.

### CODE REVIEW

**Comment by Bugagazavr** [comment]
👍 Good job

**Comment by bradrydzewski** [comment]
thanks for this!  so it sounds like status API is broken on the gitlab end? can you provide some more details so we can document as a known issue and maybe link to the gitlab issue?

**Comment by vanadium23** [comment]
@bradrydzewski don't know what is wrong. According to [gitlab docs](https://docs.gitlab.com/ce/api/commits.html#post-the-build-status-to-a-commit), the implementation is correct. Also I've rebased code onto master. I can deal with commit status in separate issue, if you want (:

**Comment by bradrydzewski** [comment]
@vanadium23 maybe check to make sure you have DRONE_HOST configured properly? If not properly configured it can impact the ability to create statuses.

**Comment by amq** [comment]
Adding the scheme to `DRONE_HOST` solved the status updates for me.

**Comment by vanadium23** [comment]
@bradrydzewski, yes, problem was in the scheme, because I set it as localhost (:

File: remote/gitlab/client/project.go [hunk 1]
```diff
@@ -1,17 +1,16 @@
 package client
 
 import (
+	b64 "encoding/base64"
```
Comment (bradrydzewski) [inline remote/gitlab/client/project.go]: remove alias and use `base64` package name below

Comment (vanadium23) [reply remote/gitlab/client/project.go]: Done.

**Comment by thomasf** [comment]
this issue closes https://github.com/drone/drone/issues/2068   (don't know why github doesnt link these issues when the issue # is in the title)

btw. Gitlab decided to keep the v3 api around a bit longerbecause they had too many v3 api requests to their site.. So latest gitlab still works with drone which is good

**Comment by vanadium23** [comment]
@bradrydzewski everything is fine?

**Comment by bradrydzewski** [comment]
ok, sorry for the delays. thanks again for this!  I plan to snapshot rc.4 on monday for those waiting for a tagged release.

**Comment by Spedge** [comment]
So we've been working on upgrading our Drone version to 0.8 - installing and delivering on rc3. rc4 is released and all of a sudden doesn't work with our supported version of GitLab? So to use Drone, we now need to do a major version upgrade of GitLab? 

As @thomasf suggests - v3 is still around. By migrating to v4 and not supporting any backward compatibility, you are really making this difficult for us.

**Comment by thomasf** [comment]
I don't think it's unfair to require the  v4 API,  the v3 API will not be maintained anymore (whatever that means). Gitlab announced that v3 was going away like a year before it happened or something and v3/v4 has been working in parallell in Gitlab for a long time giving people time to migrate.

**Comment by Spedge** [comment]
If we wrote some code to allow us to add a flag to the startup in order to use the v3 version, would you accept it?

**Comment by bradrydzewski** [comment]
@Spedge no worries. I just merged https://github.com/drone/drone/pull/2215

I ended up just copying the old gitlab code into a gitlab3 folder so that we have old and new code in sibling packages. This way I didn't have to make any major code changes in drone.

```diff
remote/gitlab
+remote/gitlab3
```

I thine added an environment variable that you can use to switch:

```
DRONE_GITLAB_V3_API=true
```

This change is available in the `drone/drone:latest` image. Would it be possible to upgrade your server and verify this change by end of day tomorrow? I only ask because I am planning on snapshotting a final release on Monday and would like to make sure there are no critical issues.

**Comment by gjtempleton** [comment]
I've just tried out the `:latest` tag with `DRONE_GITLAB_V3_API=true
` set and successfully triggered builds and had the status reported back to Gitlab successfully. Thanks for the swift response.

**Comment by bradrydzewski** [comment]
Thanks Guy for the quick reply, happy to help!

**Comment by Spedge** [comment]
Brilliant, thanks Brad - really, really appreciate the quick turnaround on this.

**Pull request merged**